examples/gcs-lock/README.md

```diff
@@ -0,0 +1,29 @@
+## GCS Lock
+
+Uses the [Google Cloud Storage generation preconditions feature](https://cloud.google.com/storage/docs/generations-preconditions#_Preconditions) 
+to create a consistent locking mechanism that can be used as a mutex. This can be used to do protected, globally 
+serialized work within a Java
+[try-with-resources](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html) block. 
+
+See [this](https://github.com/marcacohen/gcslock) GitHub repository for a [Go](https://golang.org/) based GCS lock.
+
```

Comment:
One of the most popular interfaces for GCS from Java is the Hadoop Connector which has a cooperitive locking feature [blog](https://cloud.google.com/blog/products/data-analytics/new-release-of-cloud-storage-connector-for-hadoop-improving-performance-throughput-and-more) and [config docs](https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/master/gcs/CONFIGURATION.md#cooperative-locking-feature-configuration). This is a way of providing "directory" level semantics for mutating GCS resources.

Please mention this as an alternative for HDFS / GCS users as it will better suit this specific use case.

Please describe the motivating use case(s) for using this example and the choice for GCS.

I think using GCS as a mutex is an interesting example but can seem like a strange choice without some more context. At first glance I find myself asking "Why not use Firestore transactions or a locks DB cloud sql instance or resdis/memorystore?"
It seems to me that any of these systems could provide lower latency for claiming / releasing the lock than uploading GCS objects (gut feeling, not something I've benchmarked).

The answer may be "customer already had a GCS bucket and didn't want to manage / provision another technology for this use case" but we should explain that in a public facing README.

Reply:
That's a great pointer to cooperative locking in the Hadoop Connector. Definitely something we'd want to encourage for those users. This example is a pessimistic lock.

We eventually want to supplement this sample to show it working from Dataflow, so it'll be a lot of connections to whatever mechanism we use for locking. Cloud SQL instances have a limited set of connections, so doesn't scale to the same degree as GCS. Firestore would be a good choice, but customer is already using GCS and not Firestore.